samza-yarn/src/main/scala/org/apache/samza/job/yarn/ClientHelper.scala

```diff
@@ -301,8 +301,8 @@ class ClientHelper(conf: Configuration) extends Logging {
   }
 
   private def isActiveApplication(applicationReport: ApplicationReport): Boolean = {
-    (Running.equals(toAppStatus(applicationReport).get)
-    || New.equals(toAppStatus(applicationReport).get))
+    val status = toAppStatus(applicationReport).get
```

Comment:
but application report does not change.. so how would toAppStatus return different values for the same applicationReport? i thought applicationReport does NOT change within the same invocation of isActiveAppliction.

Reply:
You are right applicationReport does not change, however, there is another check in ClientHelper#allContainersRunning to check metrics against AM on "needed-containers"

if needed-container is non zero, the status will be be NEW
only when needed-containers is zero, the status will be returned as RUNNING.

See

```
def allContainersRunning(applicationReport: ApplicationReport): Boolean = {
    val amClient: ApplicationMasterRestClient = createAmClient(applicationReport)

    debug("Created client: " + amClient.toString)

    try {
      val metrics = amClient.getMetrics
      debug("Got metrics: " + metrics.toString)

      val neededContainers = Integer.parseInt(
        metrics.get(classOf[SamzaAppMasterMetrics].getCanonicalName)
          .get("needed-containers")
          .toString)

      info("Needed containers: " + neededContainers)
      if (neededContainers == 0) {
        true
      } else {
        false
      }
```

and 
ClientHelper#toAppStatus
```      
case (YarnApplicationState.RUNNING, _) =>
        if (allContainersRunning(applicationReport)) {
          Some(Running)
        } else {
          Some(New)
        }
```

Reply:
thanks for clarifying. makes sense